lib/improver/metadata/check_datatypes.py

```diff
@@ -0,0 +1,222 @@
+#   and/or other materials provided with the distribution.
+#
+# * Neither the name of the copyright holder nor the names of its
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Utilities for datatype checking"""
+
+import numpy as np
+from cf_units import Unit
+from iris.exceptions import CoordinateNotFoundError
+
+from improver.metadata.constants.datatypes import (
+    INTEGER_QUANTITIES, TIME_METADATA)
+
+
+def check_cube_not_float64(cube, fix=False):
+    """Check a cube does not contain any float64 data, excepting time
+    coordinates. The cube can be modified in place, if the fix keyword is
+    specified to be True.
+
+    Args:
+        cube (iris.cube.Cube):
+            The input cube that will be checked for float64 inclusion.
+        fix (bool):
+            If fix is True, then the cube is amended to not include float64
+            data, otherwise, an error will be raised if float64 data is found.
+
+    Raises:
+        TypeError : Raised if float64 values are found in the cube.
+
+    """
+    if cube.dtype == np.float64:
+        if fix:
+            cube.data = cube.data.astype(np.float32)
+        else:
+            raise TypeError("64 bit cube not allowed: {!r}".format(cube))
+    for coord in cube.coords():
+        if coord.name() in ["time", "forecast_reference_time"]:
+            continue
+        if coord.points.dtype == np.float64:
+            if fix:
+                coord.points = coord.points.astype(np.float32)
+            else:
+                raise TypeError(
+                    "64 bit coord points not allowed: {} in {!r}".format(
+                        coord, cube))
+        if (hasattr(coord, "bounds") and coord.bounds is not None and
+                coord.bounds.dtype == np.float64):
+            if fix:
+                coord.bounds = coord.bounds.astype(np.float32)
+            else:
+                raise TypeError(
+                    "64 bit coord bounds not allowed: {} in {!r}".format(
+                        coord, cube))
+
+
+def check_time_coordinate_metadata(cube):
```

Comment:
Because for time coordinates (and only time coordinates) we want to check the units as well.

I had a brief discussion about this with @fionaRust and @MoseleyS .  The dilemma is that we want to check datatypes on all diagnostics, but both datatypes and units for time coordinates.  I struggled with a logical, maintainable way to split this functionality; my reasoning is:
1. Time coordinate metadata is a self-contained thing that we might like to check in a number of places.  Hence one function for this.
1. "Datatypes" is a complete standard which I'd like to treat in one function.
1. But calling both from "save" means you end up checking time coordinate datatypes twice.

This could be done without double-checking if I made the "save" checking a class or function with private members; however I think it's useful to have the two separate utilities as described.